content/reference/dos-mitigation.md

```diff
@@ -0,0 +1,311 @@
+fewer resources to execute than the damage it does. In other words, if the
+payoff is higher than the investment it is a viable attack and should be
+mitigated. Here are a couple examples
+
+1. One server opening many connections to a remote server and forcing that
+   server to spend 10x the compute time to handle the request relative to the
+   attacker server. This is attack viable because a single server amplifies it's
+   affect 10x. This attack will continue to scale if the attacker adds more
+   servers.
+
+2. 100 servers asking a single server to do some work, but if this single server
+   goes down it will indirectly cause the loss of an asset. If the asset is more
+   valuable than the compute time of 100 servers, this attack is viable.
+
+3. Many servers connecting to a single server such that that server can no
+   longer accept new connections from an honest peer. This server is now
+   isolated from the honest peers in the network. This is commonly called an
+   eclipse attack and is viable if it's either cheap to eclipse this node, or if
+   eclipsing this node has a high payoff.
+
+Generally the effect on our application can range from crashing to stalling to
+failing to handle new peers to degraded performance. Ideally we want
+our application to at worst suffer a slight perfomance penalty, but otherwise
+stay up and healthy.
+
+In the next section we'll cover some design strategies you should incorporate
+into your protocol to make sure your application stays up and healthy.
+
+# Incoporating DOS mitigation from the start
+
+The general strategy is to use the minimum amount of resources as possible, and
+make sure that there's no untrusted amplification mechanism (e.g. an untrusted
+node can force you do to 10x the work it does). A protocol level reputation
+system can help (take a look at [GossipSub](https://github.com/libp2p/specs/tree/master/pubsub/gossipsub) for inspiration) as well as
+logging misbehaving nodes and actioning those logs separately (see fail2ban
+below).
+
+Here are some more specific recommendations
+
+## Limit the number of concurrent streams per connection your protocol needs
+
+Each stream has some resource cost associated with it. Depending on the
+transport and multiplexer, this can be bigger or smaller. Try to avoid having
+too many concurrent streams open per peer for your protocol. Instead try to
+limit the maximum number of concurrent streams to something reasonable (surely
+you don't need >512 streams open at once for a peer?). Multiple concurrent
+streams can be useful for logic or to avoid [Head-of-line
+blocking](https://en.wikipedia.org/wiki/Head-of-line_blocking), but having too
+many streams will offset these benefits.
+
+Using a stream for a short period of time and then closing it is fine. It's
+really the number of _concurrent_ streams that you need to be careful of.
+
+## Limit the number of connections your application needs
```

Comment:
Given connections happen before streams in an application's lifecycle, maybe move this above the section above?